File: src/stride_core/running_calibration/zones.py
"""Training-zone generation from running calibration snapshots."""

from __future__ import annotations

from .types import (
    CalibrationConfidence,
    HeartRateZone,
    PaceZone,
    RunningCalibrationSnapshot,
    RunningZoneSet,
)

ZONE_NAMES = ("recovery", "easy", "marathon", "threshold", "interval", "repetition")

# Expressed as speed ranges relative to threshold speed. Pace bounds are the
# inverse of these speed bounds, so faster zones have smaller s/km values.
PACE_ZONE_SPEED_RATIOS: dict[str, tuple[float | None, float | None]] = {
    "recovery": (None, 0.72),
    "easy": (0.72, 0.84),
    "marathon": (0.84, 0.97),
    "threshold": (0.97, 1.03),
    "interval": (1.03, 1.11),
    "repetition": (1.11, None),
}

HR_ZONE_RATIOS: dict[str, tuple[float | None, float | None]] = {
    "recovery": (None, 0.80),
    "easy": (0.80, 0.88),
    "marathon": (0.88, 0.94),
    "threshold": (0.94, 1.01),
    "interval": (1.01, 1.06),
    "repetition": (1.06, None),
}


def _pace_s_per_km(speed_mps: float | None) -> float | None:
    if speed_mps is None or speed_mps <= 0:
        return None
    return 1000.0 / speed_mps


def compute_training_zones(snapshot: RunningCalibrationSnapshot) -> RunningZoneSet:
    pace_zones: list[PaceZone] = []
    if snapshot.threshold_speed_mps is not None and snapshot.threshold_speed_mps > 0:
        threshold = float(snapshot.threshold_speed_mps)
        for name in ZONE_NAMES:
            low_ratio, high_ratio = PACE_ZONE_SPEED_RATIOS[name]
            low_speed = threshold * low_ratio if low_ratio is not None else None
            high_speed = threshold * high_ratio if high_ratio is not None else None
            pace_zones.append(
                PaceZone(
                    name=name,
                    min_pace_s_per_km=_pace_s_per_km(high_speed),
                    max_pace_s_per_km=_pace_s_per_km(low_speed),
                    min_speed_mps=low_speed,
                    max_speed_mps=high_speed,
                    confidence=snapshot.threshold_speed_confidence,
                )
            )

    hr_zones: list[HeartRateZone] = []
    if snapshot.threshold_hr is not None and snapshot.threshold_hr > 0:
        threshold_hr = float(snapshot.threshold_hr)
        for name in ZONE_NAMES:
            low_ratio, high_ratio = HR_ZONE_RATIOS[name]
            hr_zones.append(
                HeartRateZone(
                    name=name,
                    min_bpm=threshold_hr * low_ratio if low_ratio is not None else None,
                    max_bpm=threshold_hr * high_ratio if high_ratio is not None else None,
                    confidence=snapshot.threshold_hr_confidence,
                )
            )
    elif snapshot.rhr_baseline is not None and snapshot.hrmax_estimate is not None:
        rhr = float(snapshot.rhr_baseline)
        hrmax = float(snapshot.hrmax_estimate)
        if hrmax > rhr:
            hrr_ranges = {
                "recovery": (0.55, 0.65),
                "easy": (0.65, 0.75),
                "marathon": (0.75, 0.82),
                "threshold": (0.82, 0.88),
                "interval": (0.88, 0.94),
                "repetition": (0.94, 1.0),
            }
            for name in ZONE_NAMES:
                low, high = hrr_ranges[name]
                hr_zones.append(
                    HeartRateZone(
                        name=name,
                        min_bpm=rhr + (hrmax - rhr) * low,
                        max_bpm=rhr + (hrmax - rhr) * high,
                        confidence=CalibrationConfidence.LOW,
                    )
                )

    return RunningZoneSet(
        as_of_date=snapshot.as_of_date,
        snapshot_id=snapshot.id,
        pace_zones=tuple(pace_zones),
        heart_rate_zones=tuple(hr_zones),
    )

File: src/go/internal/compute/calibration/zones.go
package calibration

// zones.go ports running_calibration/zones.py: training zones derived from a
// snapshot's threshold speed / HR (with an HRR fallback).

// PaceZone mirrors types.PaceZone.
type PaceZone struct {
	Name          string
	MinPaceSPerKm *float64
	MaxPaceSPerKm *float64
	MinSpeedMps   *float64
	MaxSpeedMps   *float64
	Confidence    Confidence
}

// HeartRateZone mirrors types.HeartRateZone.
type HeartRateZone struct {
	Name       string
	MinBpm     *float64
	MaxBpm     *float64
	Confidence Confidence
}

// ZoneSet mirrors types.RunningZoneSet (sans as_of/snapshot id, added at persist).
type ZoneSet struct {
	PaceZones      []PaceZone
	HeartRateZones []HeartRateZone
}

var zoneNames = []string{"recovery", "easy", "marathon", "threshold", "interval", "repetition"}

type ratio struct{ low, high *float64 }

func rp(v float64) *float64 { return &v }

var paceZoneSpeedRatios = map[string]ratio{
	"recovery":   {nil, rp(0.72)},
	"easy":       {rp(0.72), rp(0.84)},
	"marathon":   {rp(0.84), rp(0.97)},
	"threshold":  {rp(0.97), rp(1.03)},
	"interval":   {rp(1.03), rp(1.11)},
	"repetition": {rp(1.11), nil},
}

var hrZoneRatios = map[string]ratio{
	"recovery":   {nil, rp(0.80)},
	"easy":       {rp(0.80), rp(0.88)},
	"marathon":   {rp(0.88), rp(0.94)},
	"threshold":  {rp(0.94), rp(1.01)},
	"interval":   {rp(1.01), rp(1.06)},
	"repetition": {rp(1.06), nil},
}

var hrrRanges = map[string][2]float64{
	"recovery":   {0.55, 0.65},
	"easy":       {0.65, 0.75},
	"marathon":   {0.75, 0.82},
	"threshold":  {0.82, 0.88},
	"interval":   {0.88, 0.94},
	"repetition": {0.94, 1.0},
}

func paceSPerKm(speedMps *float64) *float64 {
	if speedMps == nil || *speedMps <= 0 {
		return nil
	}
	v := 1000.0 / *speedMps
	return &v
}

// ComputeTrainingZones mirrors zones.compute_training_zones.
func ComputeTrainingZones(snap Snapshot) ZoneSet {
	var pace []PaceZone
	if snap.ThresholdSpeedMps != nil && *snap.ThresholdSpeedMps > 0 {
		threshold := *snap.ThresholdSpeedMps
		for _, name := range zoneNames {
			r := paceZoneSpeedRatios[name]
			var lowSpeed, highSpeed *float64
			if r.low != nil {
				v := threshold * *r.low
				lowSpeed = &v
			}
			if r.high != nil {
				v := threshold * *r.high
				highSpeed = &v
			}
			pace = append(pace, PaceZone{
				Name:          name,
				MinPaceSPerKm: paceSPerKm(highSpeed),
				MaxPaceSPerKm: paceSPerKm(lowSpeed),
				MinSpeedMps:   lowSpeed,
				MaxSpeedMps:   highSpeed,
				Confidence:    snap.ThresholdSpeedConfidence,
			})
		}
	}

	var hr []HeartRateZone
	switch {
	case snap.ThresholdHR != nil && *snap.ThresholdHR > 0:
		thr := *snap.ThresholdHR
		for _, name := range zoneNames {
			r := hrZoneRatios[name]
			var minBpm, maxBpm *float64
			if r.low != nil {
				v := thr * *r.low
				minBpm = &v
			}
			if r.high != nil {
				v := thr * *r.high
				maxBpm = &v
			}
			hr = append(hr, HeartRateZone{Name: name, MinBpm: minBpm, MaxBpm: maxBpm, Confidence: snap.ThresholdHRConfidence})
		}
	case snap.RHRBaseline != nil && snap.HRMaxEstimate != nil:
		rhr := *snap.RHRBaseline
		hrmax := *snap.HRMaxEstimate
		if hrmax > rhr {
			for _, name := range zoneNames {
				rg := hrrRanges[name]
				minBpm := rhr + (hrmax-rhr)*rg[0]
				maxBpm := rhr + (hrmax-rhr)*rg[1]
				hr = append(hr, HeartRateZone{Name: name, MinBpm: &minBpm, MaxBpm: &maxBpm, Confidence: ConfidenceLow})
			}
		}
	}

	return ZoneSet{PaceZones: pace, HeartRateZones: hr}
}
